src/Compilers/CSharp/Portable/Lowering/RuntimeAsyncIteratorRewriter/RuntimeAsyncIteratorRewriter.cs

```diff
@@ -0,0 +1,428 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.Emit;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp;
+
+/// <summary>
+/// The entrypoint for this rewriter is <see cref="Rewrite"/>.
+/// It delegates to <see cref="StateMachineRewriter.Rewrite"/> which drives the process relying on various overrides
+/// to produce a nested type <see cref="RuntimeAsyncIteratorStateMachine"/> with:
+/// - fields:
+///   - <see cref="GenerateControlFields"/> for `current`, `disposeMode`, `initialThreadId` fields
+///   - proxies for `this`, parameters and hoisted locals
+/// - members including interface implementations:
+///   - <see cref="GenerateConstructor"/>
+///   - <see cref="GenerateIAsyncEnumerableImplementation_GetAsyncEnumerator"/> (for enumerable case)
+///   - <see cref="GenerateIAsyncEnumeratorImplementation_MoveNextAsync"/> which relies heavily on <see cref="MoveNextAsyncRewriter"/>
+///   - <see cref="GenerateIAsyncEnumeratorImplementation_Current"/>
+///   - <see cref="GenerateIAsyncDisposable_DisposeAsync"/>
+/// before returning the body for the kickoff method (<see cref="StateMachineRewriter.GenerateKickoffMethodBody"/>).
+/// 
+/// The state machines uses the same states as regular async-iterator methods, with the exception of `await` states
+/// which are no longer needed (handled by the runtime directly).
+/// </summary>
+internal sealed partial class RuntimeAsyncIteratorRewriter : StateMachineRewriter
+{
+    // true if the iterator implements IAsyncEnumerable<T>,
+    // false if it implements IAsyncEnumerator<T>
```

Comment:
Nit: the wording implies that it won't implement `IAsyncEnumerator<T>` when `true`, which isn't true, right?